force-app/main/default/classes/OpportunityTriggerHandler.cls

```diff
@@ -0,0 +1,186 @@
+public class OpportunityTriggerHandler extends TriggerHandler {
+  public override void beforeInsert() {
+    // Set default Type for new Opportunities
+    for (Opportunity opp : (List<Opportunity>) Trigger.new) {
+      if (opp.Type == null) {
+        opp.Type = 'New Customer';
+      }
+    }
+  }
+
+  public override void beforeUpdate() {
+    // Validate that the amount is greater than 5000
+    for (Opportunity opp : (List<Opportunity>) Trigger.new) {
+      if (opp.Amount < 5000) {
+        opp.addError('Opportunity amount must be greater than 5000');
+      }
+    }
+
+    // Set the primary contact to the CEO if the field is empty
+    Set<Id> accountIds = new Set<Id>();
+    for (Opportunity opp : (List<Opportunity>) Trigger.new) {
+      if (opp.AccountId != null) {
+        accountIds.add(opp.AccountId);
+      }
+    }
+
+    if (!accountIds.isEmpty()) {
+      Map<Id, Contact> ceoContacts = new Map<Id, Contact>(
+        [
+          SELECT Id, AccountId
+          FROM Contact
+          WHERE AccountId IN :accountIds AND Title = 'CEO'
+        ]
+      );
+
+      for (Opportunity opp : (List<Opportunity>) Trigger.new) {
+        if (
+          opp.Primary_Contact__c == null &&
+          ceoContacts.containsKey(opp.AccountId)
+        ) {
+          opp.Primary_Contact__c = ceoContacts.get(opp.AccountId).Id;
+        }
+      }
+    }
+  }
+
+  public override void afterInsert() {
+    // Create a new Task for newly inserted Opportunities with a Primary Contact
+    List<Task> tasksToInsert = new List<Task>();
+    for (Opportunity opp : (List<Opportunity>) Trigger.new) {
+      if (opp.Primary_Contact__c != null) {
+        Task tsk = new Task(
+          Subject = 'Call Primary Contact',
+          WhatId = opp.Id,
+          WhoId = opp.Primary_Contact__c,
+          OwnerId = opp.OwnerId,
+          ActivityDate = Date.today().addDays(3)
+        );
+        tasksToInsert.add(tsk);
+      }
+    }
+    if (!tasksToInsert.isEmpty()) {
+      insert tasksToInsert;
+    }
+  }
+
+  public override void afterUpdate() {
+    // Append Stage changes to the Opportunity Description
+    List<Opportunity> opportunitiesToUpdate = new List<Opportunity>();
+    for (Integer i = 0; i < Trigger.new.size(); i++) {
+      Opportunity newOpp = (Opportunity) Trigger.new[i];
+      Opportunity oldOpp = (Opportunity) Trigger.old[i];
+      if (newOpp.StageName != oldOpp.StageName && newOpp.StageName != null) {
+        newOpp.Description +=
+          '\n Stage Change: ' +
+          newOpp.StageName +
+          ' : ' +
+          DateTime.now().format();
+        opportunitiesToUpdate.add(newOpp);
+      }
+    }
+    if (!opportunitiesToUpdate.isEmpty()) {
+      update opportunitiesToUpdate;
```

Comment:
What happens when you update the same records after they were just updated? Can you prevent this recursion?